src/offset/local/unix.rs

```diff
@@ -8,64 +8,118 @@
+                fs::symlink_metadata("/etc/localtime")
+                    .expect("localtime should exist")
+                    .modified()
+                    .unwrap(),
             ),
-            LocalResult::Single(offset) => {
-                LocalResult::Single(DateTime::from_utc(*d - offset, offset))
-            }
         }
-    } else {
-        LocalResult::Single(DateTime::from_utc(*d, offset(*d, false).unwrap()))
     }
 }
 
-fn offset(d: NaiveDateTime, local: bool) -> LocalResult<FixedOffset> {
-    let info = unsafe {
-        INIT.call_once(|| {
-            INFO = Some(TimeZone::local().expect("unable to parse localtime info"));
-        });
-        INFO.as_ref().unwrap()
-    };
+impl Source {
+    fn out_of_date(&self) -> bool {
+        let current = Source::default();
 
-    if local {
-        // we pass through the year as the year of a local point in time must either be valid in that locale, or
-        // the entire time was skipped in which case we will return LocalResult::None anywa.
-        match info
-            .find_local_time_type_from_local(d.timestamp(), d.year())
-            .expect("unable to select local time type")
-        {
-            LocalResult::None => LocalResult::None,
-            LocalResult::Ambiguous(early, late) => LocalResult::Ambiguous(
-                FixedOffset::east(early.offset()),
-                FixedOffset::east(late.offset()),
-            ),
-            LocalResult::Single(tt) => LocalResult::Single(FixedOffset::east(tt.offset())),
+        // can remove the `ref` if/when MSRV increases to allow mixed owned/ref patterns
+        match (self, current) {
+            (Source::LocalTime(prev), Source::LocalTime(ref now)) => prev == now,
+            (Source::Environment(prev), Source::Environment(ref now)) => prev == now,
+            _ => false,
+        }
+    }
+}
+
+struct Cache {
+    zone: TimeZone,
+    source: Source,
+}
+
+impl Default for Cache {
+    fn default() -> Cache {
+        Cache {
+            zone: TimeZone::local().expect("unable to parse localtime info"),
+            source: Source::default(),
         }
-    } else {
-        LocalResult::Single(FixedOffset::east(
-            info.find_local_time_type(d.timestamp())
-                .expect("unable to select local time type")
-                .offset(),
-        ))
     }
 }
 
-static mut INFO: Option<TimeZone> = None;
-static INIT: Once = Once::new();
+impl Cache {
+    fn get_offset(&mut self, d: NaiveDateTime, local: bool) -> LocalResult<DateTime<Local>> {
+        if self.source.out_of_date() {
+            *self = Cache::default();
+        }
+
+        if !local {
+            let offset = FixedOffset::east(
+                self.zone
+                    .find_local_time_type(d.timestamp())
+                    .expect("unable to select local time type")
+                    .offset(),
+            );
+
+            LocalResult::Single(DateTime::from_utc(d, offset))
```

Comment:
Can we just return here so we don't need the extra level of indentation for the rest of the function?